internal/ais/workflow.go

```diff
@@ -0,0 +1,222 @@
+package ais
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"path/filepath"
+	"time"
+
+	"github.com/artefactual-sdps/temporal-activities/archivezip"
+	"github.com/artefactual-sdps/temporal-activities/bucketupload"
+	"github.com/artefactual-sdps/temporal-activities/removepaths"
+	"github.com/google/uuid"
+	"go.artefactual.dev/tools/bucket"
+	temporalsdk_api_enums "go.temporal.io/api/enums/v1"
+	temporalsdk_activity "go.temporal.io/sdk/activity"
+	temporalsdk_client "go.temporal.io/sdk/client"
+	temporalsdk_temporal "go.temporal.io/sdk/temporal"
+	temporalsdk_worker "go.temporal.io/sdk/worker"
+	temporalsdk_workflow "go.temporal.io/sdk/workflow"
+)
+
+const WorkflowName = "ais-workflow"
+
+type WorkflowParams struct {
+	AIPUUID uuid.UUID
+	AIPName string
+}
+
+type WorkflowResult struct {
+	Key string
+}
+
+type Workflow struct {
+	workingDir string
+}
+
+func NewWorkflow(workingDir string) *Workflow {
+	return &Workflow{workingDir: workingDir}
+}
+
+func (w *Workflow) Execute(ctx temporalsdk_workflow.Context, params *WorkflowParams) (r *WorkflowResult, e error) {
+	r = &WorkflowResult{}
+	removePaths := []string{}
+
+	logger := temporalsdk_workflow.GetLogger(ctx)
+	logger.Debug("AIS workflow running!", "params", params)
+
+	defer func() {
+		var removeResult removepaths.Result
+		err := temporalsdk_workflow.ExecuteActivity(
+			withLocalActOpts(ctx),
+			removepaths.Name,
+			&removepaths.Params{Paths: removePaths},
+		).Get(ctx, &removeResult)
+		if err != nil {
+			e = errors.Join(e, err)
+		}
+
+		logger.Debug("AIS workflow finished!", "result", r, "error", e)
+	}()
+
+	aipDirName := fmt.Sprintf("%s-%s", params.AIPName, params.AIPUUID.String())
+	localDir := filepath.Join(w.workingDir, fmt.Sprintf("search-md_%s", aipDirName))
+	metsName := fmt.Sprintf("METS.%s.xml", params.AIPUUID.String())
+	metsPath := filepath.Join(localDir, metsName)
+
+	removePaths = append(removePaths, localDir)
+
+	var fetchMETSResult FetchActivityResult
+	e = temporalsdk_workflow.ExecuteActivity(
+		withRemoteActOpts(ctx),
```

Comment:
I don't think the fetch activity is sending a heartbeat, so it will timeout if the download time exceeds the heartbeat timeout of one minute.

Reply:
Good catch, thanks!